### PR TITLE
Nvidia runtime

### PR DESCRIPTION
**Pull request by oleksii-shyman**
Add support for "nvidia" runtime for "docker" driver

TO BE DONE:

1. Rebase commits before merging (get rid of the latest commit)

### CODE REVIEW

File: drivers/docker/driver.go [hunk 1]
```diff
@@ -624,6 +625,10 @@ func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *T
 		PidsLimit: driverConfig.PidsLimit,
 	}
 
+	if _, ok := task.Env[nvidia.NvidiaVisibleDevices]; ok {
+		hostConfig.Runtime = "nvidia"
```
Comment (oleksii-shyman) [inline drivers/docker/driver.go]: proper place for storing "nvidia" runtime name constant should be defined

File: drivers/docker/fingerprint.go [hunk 2]
```diff
@@ -88,5 +89,17 @@ func (d *Driver) buildFingerprint() *drivers.Fingerprint {
...
 
+	if dockerInfo, err := client.Info(); err != nil {
+		d.logger.Warn("failed to get docker system info", "error", err)
+	} else {
+		runtimeNames := make([]string, 0, len(dockerInfo.Runtimes))
+		for name := range dockerInfo.Runtimes {
+			runtimeNames = append(runtimeNames, name)
+		}
+
+		fp.Attributes["Runtimes"] = pstructs.NewStringAttribute(
```
Comment (oleksii-shyman) [inline drivers/docker/fingerprint.go]: The proper place for storing "Runtimes" key should be defined

File: drivers/docker/fingerprint.go [hunk 3]
```diff
@@ -88,5 +89,20 @@ func (d *Driver) buildFingerprint() *drivers.Fingerprint {
...
+	} else {
+		runtimeNames := make([]string, 0, len(dockerInfo.Runtimes))
+		for name := range dockerInfo.Runtimes {
+			if d.config.GPURuntimeName == name {
+				d.gpuRuntime = true
+			}
+			runtimeNames = append(runtimeNames, name)
+		}
+
+		fp.Attributes["runtimes"] = pstructs.NewStringAttribute(
```
Comment (oleksii-shyman) [inline drivers/docker/fingerprint.go]: technically, "runtimes" attribute is not used, and can be deleted.
Should we keep it?

File: drivers/docker/config.go [hunk 4]
```diff
@@ -153,6 +159,7 @@ var (
 	//		}
 	//		allow_privileged = false
 	//		allow_caps = ["CHOWN", "NET_RAW" ... ]
+	//      nvidia_runtime = "nvidia"
```
Comment (dadgar) [inline drivers/docker/config.go]: Fix spacing

File: drivers/docker/driver.go [hunk 5]
```diff
@@ -628,6 +632,13 @@ func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *T
 		PidsLimit: driverConfig.PidsLimit,
 	}
 
+	if _, ok := task.DeviceEnv[nvidia.NvidiaVisibleDevices]; ok {
+		if !d.gpuRuntime {
+			return c, fmt.Errorf("requested docker-runtime '%s' was not found", d.config.GPURuntimeName)
```
Comment (dadgar) [inline drivers/docker/driver.go]: '%s' -> %q

File: drivers/docker/driver_test.go [hunk 6]
```diff
@@ -1168,6 +1170,80 @@ func TestDockerDriver_CreateContainerConfig(t *testing.T) {
...
+	if !tu.IsTravis() {
+		t.Parallel()
+	}
+	if !testutil.DockerIsConnected(t) {
+		t.Skip("Docker not connected")
+	}
+	if runtime.GOOS != "linux" {
+		t.Skip("nvidia plugin supports only linux")
+	}
+	for _, testCase := range []struct {
```
Comment (dadgar) [inline drivers/docker/driver_test.go]: Define the cases outside the for-loop and use t.Run():
https://github.com/hashicorp/nomad/blob/6fb074e0e8a762d9d4fe2898233ad49bd59a0430/plugins/shared/structs/attribute_test.go#L12-L75

File: drivers/docker/fingerprint.go [hunk 7]
```diff
@@ -88,5 +89,20 @@ func (d *Driver) buildFingerprint() *drivers.Fingerprint {
 		}
 	}
 
+	if dockerInfo, err := client.Info(); err != nil {
+		d.logger.Warn("failed to get docker system info", "error", err)
```
Comment (dadgar) [inline drivers/docker/fingerprint.go]: Capitalize Docker

File: drivers/docker/fingerprint.go [hunk 8]
```diff
@@ -88,5 +89,20 @@ func (d *Driver) buildFingerprint() *drivers.Fingerprint {
...
+	} else {
+		runtimeNames := make([]string, 0, len(dockerInfo.Runtimes))
+		for name := range dockerInfo.Runtimes {
+			if d.config.GPURuntimeName == name {
+				d.gpuRuntime = true
+			}
+			runtimeNames = append(runtimeNames, name)
```
Comment (dadgar) [inline drivers/docker/fingerprint.go]: Sort these at the end so we have guaranteed ordering

File: drivers/docker/fingerprint.go (hunk 3 above)
Comment (dadgar) [reply drivers/docker/fingerprint.go]: I would keep it. Could be useful for constraints

File: drivers/docker/fingerprint.go [hunk 9]
```diff
@@ -88,5 +89,20 @@ func (d *Driver) buildFingerprint() *drivers.Fingerprint {
 		}
 	}
 
+	if dockerInfo, err := client.Info(); err != nil {
+		d.logger.Warn("failed to get docker system info", "error", err)
+	} else {
+		runtimeNames := make([]string, 0, len(dockerInfo.Runtimes))
+		for name := range dockerInfo.Runtimes {
+			if d.config.GPURuntimeName == name {
```
Comment (dadgar) [inline drivers/docker/fingerprint.go]: Add a comment explaining this

File: website/source/docs/drivers/docker.html.md [hunk 10]
```diff
@@ -688,6 +688,9 @@ options](/docs/configuration/client.html#options):
   disable Nomad from removing a container when the task exits. Under a name
   conflict, Nomad may still remove the dead container.
 
+* `docker.nvidia_runtime`: Defaults to `nvidia`. This option can be used to
```
Comment (dadgar) [inline website/source/docs/drivers/docker.html.md]: This option allows operators to select the runtime that should be used in order to expose Nvidia GPUs to the container.

**Pull request merged**